### ISSUE TITLE
new Check: single line annotation location

### ISSUE THREAD

**Issue by vanniktech** [checkstyle/checkstyle#3440]
I'd love to have [AnnotationLocation](http://checkstyle.sourceforge.net/config_annotation.html) extended in a way that the following code would fail:

``` java
@NonNull
private List<String> names = new ArrayList<>();
```

Since it should be 

``` java
@NonNull private List<String> names = new ArrayList<>();
```

Until version 7.1.1 this is not possible. You can only white list the latter sample but not tell the check to fail on the first example.

update:
AnnotationLocation can not be updated as it have 3 properties that allowSamelineXXXXXX . So update have new option that will demand singleline will looks weird in Checkdesign. So new check have to be created.
##

**Comment by romani** [comment]
> Since it should be

why ? what rules you propose to make it happen ?

Please provide a config that you used, this check have a lot of properties.

attention: beware that we can not distinguish target of annotation (type or field itself).

**Comment by vanniktech** [comment]
Basically another configuration that enforces that annotations need to be on the same line.

**Comment by romani** [comment]
Do you want all annotations for all fields to be on the same line ?
You have minimalistic code example some fields have a lot of annotations (for example for hibernate mapping).

**Comment by vanniktech** [comment]
I'd say so yes that all annotations should be on the same line.

**Comment by romani** [comment]
@vanniktech ,

Do you expect violations on annotation locations like:
1)
```
@SuppressWarnings("deprecation")
@Mock DataLoader loader;
```
2)
```
@SuppressWarnings("deprecation")
@Mock
DataLoader loader;
```

and NO violations on code like:
3)
```
@SuppressWarnings("deprecation") @Mock @Mock2 @Mock3 @Mock4 DataLoader loader;
```

please provide more cases with code and notes where you expect violations.
As forcing all annotations to be on single line is not always practical. From my perspective only single non-parameters annotation is ok to keep on the same line.

**Comment by vanniktech** [comment]
### First test case

#### Input 

```java
@SuppressWarnings("deprecation")
@Mock DataLoader loader;
```

#### Output

1 violation since I'd like to be like this:

```java
@SuppressWarnings("deprecation") @Mock DataLoader loader;
```

### Second

#### Input 

```java
@SuppressWarnings("deprecation")
@Mock
DataLoader loader;

#### Output

```
2 violations (one per annotation) since I'd like to be like this:

```java
@SuppressWarnings("deprecation") @Mock DataLoader loader;
```

### Third

#### Input 

```java
@SuppressWarnings("deprecation") @Mock @Mock2 @Mock3 @Mock4 DataLoader loader;
```

#### Output

0 violations since all annotations are aligned properly.

---

> As forcing all annotations to be on single line is not always practical. From my perspective only single non-parameters annotation is ok to keep on the same line.

Maybe it's possible to use the MaxLineLength property for this and allow it to break if the limit is reached? In 99% of the cases they don't exceed the line limit. Even `@SuppressWarnings("deprecation") @Mock @Mock2 @Mock3 @Mock4 DataLoader loader;` is still fine since it has 78 characters.

I hope this is clear enough.

**Comment by romani** [comment]
> Maybe it's possible to use the MaxLineLength property for this and allow it to break if the limit is reached?

we can not use other Check settings, all Checks are independent. You will have to play with suppression if you want to avoid conflicts.

**Comment by vanniktech** [comment]
Then make it a property of this new check and give it the same default value?

**Comment by romani** [comment]
no, it is sign of bad design or bad idea of Check.
If user come to extreme code that is still valid from his perspective  - he need to use suppressions.

**Comment by romani** [comment]
fix is merged.
